File: go/019.go
package main

/**
给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。
思路：倒数 n 个，先求总长度，然后找到idx
idx = len - n + 1  *注意删除头节点的情况 n == len head = head.next
*/
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	node, len, idx := head, 0, 0
	for node != nil {
		len++
		node = node.Next
	}
	p := head
	if p.Next == nil {
		return nil
	}
	node, idx = head.Next, len-n+1
	if n == len {
		return head.Next
	}
	for idx > 2 {
		node = node.Next
		p = p.Next
		idx--
	}
	p.Next = node.Next
	return head
}

File: c++/019.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *p,*q,*pre;
        int cn=0;
        p=q=head;
        while(q){
            if(cn>=n){
                pre = p;
                p = p->next;
            }
            q = q->next;
            cn++;
        }
        if(cn==n){
            p = head;
            head = head->next;
            delete p;
        }else if(cn>n){
            pre->next = p->next;
            delete p;
        }
        return head;
    }
};